pkg/cache/v2/simple.go

```diff
@@ -94,11 +100,56 @@ type snapshotCache struct {
 // Logger is optional.
 func NewSnapshotCache(ads bool, hash NodeHash, logger log.Logger) SnapshotCache {
 	return &snapshotCache{
-		log:       logger,
-		ads:       ads,
-		snapshots: make(map[string]Snapshot),
-		status:    make(map[string]*statusInfo),
-		hash:      hash,
+		log:               logger,
+		ads:               ads,
+		snapshots:         make(map[string]Snapshot),
+		status:            make(map[string]*statusInfo),
+		hash:              hash,
+		heartbeatRoutines: make(map[string]heartbeatHandle),
+	}
+}
+
+func (cache *snapshotCache) sendHeartbeats(ctx context.Context, wg *sync.WaitGroup, node string) {
+	snapshot := cache.snapshots[node]
+	if info, ok := cache.status[node]; ok {
+		info.mu.Lock()
+		for id, watch := range info.watches {
+			// Respond with the current version regardless of whether the version has changed.
+			version := snapshot.GetVersion(watch.Request.TypeUrl)
+			resources := snapshot.GetResourcesAndTtl(watch.Request.TypeUrl)
+
+			// TODO(snowp): Construct this once per type instead of once per watch.
+			resourcesWithTtl := map[string]types.ResourceWithTtl{}
+			for k, v := range resources {
+				if v.Ttl != nil {
+					resourcesWithTtl[k] = v
+				}
+			}
+
+			if len(resourcesWithTtl) == 0 {
+				continue
+			}
+			if cache.log != nil {
+				cache.log.Debugf("respond open watch %d%v with heartbeat for version %q", id, watch.Request.ResourceNames, version)
+			}
+
+			wg.Add(1)
+			// Spawn a goroutine here so that we can give up the lock before issuing responses. This prevents lock contention if the
+			// response channels aren't ready to receive the response.
+			go func(request *Request, response chan<- Response) {
+				defer wg.Done()
+				select {
+				case <-ctx.Done():
+					return
+				default:
+				}
+				cache.respond(request, response, resourcesWithTtl, version, true)
```

Comment:
sending hearbeatg is essentially sending an actual response and not a simple healthcheck style check. In a highly concurrent system, can this out of band send cause race condition bugs where the heartbeat code path overrides a more recent response? Also, what happens if the response is lost for any reason(flow control, etc) ?

Reply:
Not sure what you mean by a health check; the heartbeats special responses that only serve to refresh the TTL timeout for the resources and will never 

I could be wrong but my understanding of the cache/server interaction here is that the responses will be serially sequenced over the streams regardless of load due to the use of channels, so I don't see how an old heartbeat response could come before a newer update. We could add a safeguard to Envoy where we discard heartbeat-like responses if the version doesn't match if that makes things more palatable. 

Does GCP ever drop responses? That seems problematic in itself for a general purpose control plane as I'm pretty sure that would violate some of the eventual consistency expectations?